README.md

BriefLZ - small fast Lempel-Ziv
===============================

Version 1.3.0

Copyright (c) 2002-2020 Joergen Ibsen

<http://www.ibsensoftware.com/>

[![Build Status](https://travis-ci.org/jibsen/brieflz.svg?branch=master)](https://travis-ci.org/jibsen/brieflz) [![Build status](https://ci.appveyor.com/api/projects/status/l9vhammx8p8hkrqb/branch/master?svg=true)](https://ci.appveyor.com/project/jibsen/brieflz/branch/master) [![codecov](https://codecov.io/gh/jibsen/brieflz/branch/master/graph/badge.svg)](https://codecov.io/gh/jibsen/brieflz)


About
-----

BriefLZ is a small and fast open source implementation of a Lempel-Ziv
style compression algorithm. The main focus is on speed and code footprint,
but the ratios achieved are quite good compared to similar algorithms.


Why BriefLZ?
------------

Two widely used types of Lempel-Ziv based compression libraries are those
that employ [entropy encoding][entropy] to achieve good ratios (like Brotli,
zlib, Zstd), and those that forgo entropy encoding to favor speed (like LZ4,
LZO).

BriefLZ attempts to place itself somewhere between these two by using a
[universal code][universal], which may improve compression ratios compared to
no entropy encoding, while requiring little extra code.

Not counting the optional lookup tables, the compression function `blz_pack`
is 147 LOC, and the decompression function `blz_depack` is 61 LOC (and can be
implemented in 103 bytes of x86 machine code).

If you do not need the extra speed of libraries without entropy encoding, but
want reasonable compression ratios, and the footprint of the compression or
decompression code is a factor, BriefLZ might be an option.

[entropy]: https://en.wikipedia.org/wiki/Entropy_encoding
[universal]: https://en.wikipedia.org/wiki/Universal_code_(data_compression)


Benchmark
---------

Here are some results from running [lzbench][] on the
[Silesia compression corpus][silesia] on a Core i5-4570 @ 3.2GHz:

| Compressor name         | Compression| Decompress.|  Compr. size  | Ratio |
| ---------------         | -----------| -----------| ------------- | ----- |
| zstd 1.3.4 -9           |    30 MB/s |   894 MB/s |    60,691,094 | 3.492 |
| brotli 2017-12-12 -4    |    44 MB/s |   383 MB/s |    64,202,945 | 3.301 |
| **brieflz 1.3.0 -6**    |    17 MB/s |   375 MB/s |    67,208,420 | 3.153 |
| zlib 1.2.11 -6          |    24 MB/s |   295 MB/s |    68,228,431 | 3.106 |
| zstd 1.3.4 -1           |   310 MB/s |   943 MB/s |    73,654,014 | 2.877 |
| **brieflz 1.3.0 -3**    |   112 MB/s |   341 MB/s |    75,550,736 | 2.805 |
| zlib 1.2.11 -1          |    68 MB/s |   279 MB/s |    77,259,029 | 2.743 |
| lz4hc 1.8.2 -9          |    29 MB/s |  2901 MB/s |    77,884,448 | 2.721 |
| brotli 2017-12-12 -0    |   258 MB/s |   297 MB/s |    78,432,913 | 2.702 |
| **brieflz 1.3.0 -1**    |   173 MB/s |   329 MB/s |    81,138,803 | 2.612 |
| lzo1x 2.09              |   498 MB/s |   617 MB/s |   100,572,537 | 2.107 |
| lz4 1.8.2               |   547 MB/s |  2955 MB/s |   100,880,800 | 2.101 |

Please note that this benchmark is not entirely fair because BriefLZ has no
window size limit.

[lzbench]: https://github.com/inikep/lzbench
[silesia]: http://sun.aei.polsl.pl/~sdeor/index.php?page=silesia


Usage
-----

The include file `include/brieflz.h` contains documentation in the form of
[doxygen][] comments. A configuration file is included, so you can simply run
`doxygen` to generate documentation in HTML format.

If you wish to compile BriefLZ on 16-bit systems, make sure to adjust the
constants `BLZ_HASH_BITS` and `DEFAULT_BLOCK_SIZE`.

When using BriefLZ as a shared library (dll on Windows), define `BLZ_DLL`.
When building BriefLZ as a shared library, define both `BLZ_DLL` and
`BLZ_DLL_EXPORTS`.

The `example` folder contains a simple command-line program, `blzpack`, that
can compress and decompress a file using BriefLZ. For convenience, the example
comes with makefiles for GCC and MSVC.

BriefLZ uses [CMake][] to generate build systems. To create one for the tools
on your platform, and build BriefLZ, use something along the lines of:

~~~sh
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Release ..
cmake --build . --config Release
~~~

You can also simply compile the source files and link them into your project.
CMake just provides an easy way to build and test across various platforms and
toolsets.

[doxygen]: http://www.doxygen.org/
[CMake]: http://www.cmake.org/


License
-------

This projected is licensed under the [zlib License](LICENSE) (Zlib).


NEWS
BriefLZ 1.3.0 (TBD)

  * Add btparse algorithm, which is used for levels 8, 9, and optimal.
  * Reduce memory usage on 64-bit Linux.

BriefLZ 1.2.0 (2018-10-25)

  * Add compression function `blz_pack_level` that optionally provides
    better ratios by using more time/space
  * Slightly improve compression and decompression speed
  * The source is using some C99 features, for OpenWatcom compile with
    `-za99` or as C++ with `-cc++`.

BriefLZ 1.1.0 (2015-09-05)

  * Remove legacy code and update build system
  * Use `BLZ_HASH_BITS` to set workmem size
  * Adjust default block size to 1024k in blzpack
  * Default to not performing slow CRC computations

BriefLZ 1.05 (unreleased)

  * Add safe decompression functions
  * Fix a potential portability issue

BriefLZ 1.04 (2004-06-30)

  * Gather all functions in one header file
  * Add pre-compiled libraries
  * Add .NET wrapper
  * Rewrite documentation

BriefLZ 1.02 (2003-08-14)

  * Make it more obvious that `blz_depack` requires the length of the
    depacked data, not the packed data
  * Update documentation
  * Update C source

BriefLZ 1.00 (2003-03-17)

  * First release

Project started November 12th 2002.


src/depacks.c
/*
 * BriefLZ - small fast Lempel-Ziv
 *
 * C safe depacker
 *
 * Copyright (c) 2002-2018 Joergen Ibsen
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 *   1. The origin of this software must not be misrepresented; you must
 *      not claim that you wrote the original software. If you use this
 *      software in a product, an acknowledgment in the product
 *      documentation would be appreciated but is not required.
 *
 *   2. Altered source versions must be plainly marked as such, and must
 *      not be misrepresented as being the original software.
 *
 *   3. This notice may not be removed or altered from any source
 *      distribution.
 */

#include "brieflz.h"

/* Internal data structure */
struct blz_state {
	const unsigned char *src;
	unsigned char *dst;
	unsigned long src_avail;
	unsigned long dst_avail;
	unsigned int tag;
	int bits_left;
};

#if !defined(BLZ_NO_LUT)
static const unsigned char blz_gamma_lookup[256][2] = {
	/* 00xxxxxx = 2 */
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},
	{2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2}, {2, 2},

	/* 0100xxxx = 4 */
	{4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4},
	{4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4}, {4, 4},

	/* 010100xx = 8 */
	{8, 6}, {8, 6}, {8, 6}, {8, 6},

	/* 01010100 = 16  01010101 = 16+ 01010110 = 17  01010111 = 17+ */
	{16, 8}, {16, 0}, {17, 8}, {17, 0},

	/* 010110xx = 9 */
	{9, 6}, {9, 6}, {9, 6}, {9, 6},

	/* 01011100 = 18  01011101 = 18+ 01011110 = 19  01011111 = 19+ */
	{18, 8}, {18, 0}, {19, 8}, {19, 0},

	/* 0110xxxx = 5 */
	{5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4},
	{5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4}, {5, 4},

	/* 011100xx = 10 */
	{10, 6}, {10, 6}, {10, 6}, {10, 6},

	/* 01110100 = 20  01110101 = 20+ 01110110 = 21  01110111 = 21+ */
	{20, 8}, {20, 0}, {21, 8}, {21, 0},

	/* 011110xx = 11 */
	{11, 6}, {11, 6}, {11, 6}, {11, 6},

	/* 01111100 = 22  01111101 = 22+ 01111110 = 23  01111111 = 23+ */
	{22, 8}, {22, 0}, {23, 8}, {23, 0},

	/* 10xxxxxx = 3 */
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},
	{3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2}, {3, 2},

	/* 1100xxxx = 6 */
	{6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4},
	{6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4}, {6, 4},

	/* 110100xx = 12 */
	{12, 6}, {12, 6}, {12, 6}, {12, 6},

	/* 11010100 = 24  11010101 = 24+ 11010110 = 25  11010111 = 25+ */
	{24, 8}, {24, 0}, {25, 8}, {25, 0},

	/* 110110xx = 13 */
	{13, 6}, {13, 6}, {13, 6}, {13, 6},

	/* 11011100 = 26  11011101 = 26+ 11011110 = 27  11011111 = 27+ */
	{26, 8}, {26, 0}, {27, 8}, {27, 0},

	/* 1110xxxx = 7 */
	{7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4},
	{7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4}, {7, 4},

	/* 111100xx = 14 */
	{14, 6}, {14, 6}, {14, 6}, {14, 6},

	/* 11110100 = 28  11110101 = 28+ 11110110 = 29  11110111 = 29+ */
	{28, 8}, {28, 0}, {29, 8}, {29, 0},

	/* 111110xx = 15 */
	{15, 6}, {15, 6}, {15, 6}, {15, 6},

	/* 11111100 = 30  11111101 = 30+ 11111110 = 31  11111111 = 31+ */
	{30, 8}, {30, 0}, {31, 8}, {31, 0}
};
#endif

static int
blz_getbit_safe(struct blz_state *bs, unsigned int *result)
{
	unsigned int bit;

	/* Check if tag is empty */
	if (!bs->bits_left--) {
		if (bs->src_avail < 2) {
			return 0;
		}
		bs->src_avail -= 2;

		/* Load next tag */
		bs->tag = (unsigned int) bs->src[0]
		       | ((unsigned int) bs->src[1] << 8);
		bs->src += 2;
		bs->bits_left = 15;
	}

	/* Shift bit out of tag */
	bit = (bs->tag & 0x8000) ? 1 : 0;
	bs->tag <<= 1;

	*result = bit;

	return 1;
}

static int
blz_getgamma_safe(struct blz_state *bs, unsigned long *result)
{
	unsigned int bit;
	unsigned long v = 1;

#if !defined(BLZ_NO_LUT)
	if (bs->bits_left >= 8) {
		unsigned int top8 = (bs->tag >> 8) & 0x00FF;
		int shift;

		v = blz_gamma_lookup[top8][0];
		shift = (int) blz_gamma_lookup[top8][1];

		if (shift) {
			bs->tag <<= shift;
			bs->bits_left -= shift;

			*result = v;

			return 1;
		}

		bs->tag <<= 8;
		bs->bits_left -= 8;
	}
#endif

	/* Input gamma2-encoded bits */
	do {
		if (!blz_getbit_safe(bs, &bit)) {
			return 0;
		}

		if (v & 0x80000000UL) {
			return 0;
		}

		v = (v << 1) + bit;

		if (!blz_getbit_safe(bs, &bit)) {
			return 0;
		}
	} while (bit);

	*result = v;

	return 1;
}

unsigned long
blz_depack_safe(const void *src, unsigned long src_size,
                void *dst, unsigned long depacked_size)
{
	struct blz_state bs;
	unsigned long dst_size = 0;
	unsigned int bit;

	bs.src = (const unsigned char *) src;
	bs.src_avail = src_size;
	bs.dst = (unsigned char *) dst;
	bs.dst_avail = depacked_size;

	/* Initialise to one bit left in tag; that bit is zero (a literal) */
	bs.bits_left = 1;
	bs.tag = 0;

	/* Main decompression loop */
	while (dst_size < depacked_size) {
		if (!blz_getbit_safe(&bs, &bit)) {
			return BLZ_ERROR;
		}

		if (bit) {
			unsigned long len;
			unsigned long off;

			/* Input match length and offset */
			if (!blz_getgamma_safe(&bs, &len)) {
				return BLZ_ERROR;
			}
			if (!blz_getgamma_safe(&bs, &off)) {
				return BLZ_ERROR;
			}

			len += 2;
			off -= 2;

			if (off >= 0x00FFFFFFUL) {
				return BLZ_ERROR;
			}

			if (!bs.src_avail--) {
				return BLZ_ERROR;
			}

			off = (off << 8) + (unsigned long) *bs.src++ + 1;

			if (off > depacked_size - bs.dst_avail) {
				return BLZ_ERROR;
			}

			if (len > bs.dst_avail) {
				return BLZ_ERROR;
			}

			bs.dst_avail -= len;

			/* Copy match */
			{
				const unsigned char *p = bs.dst - off;
				unsigned long i;

				for (i = len; i > 0; --i) {
					*bs.dst++ = *p++;
				}
			}

			dst_size += len;
		}
		else {
			/* Copy literal */
			if (!bs.src_avail-- || !bs.dst_avail--) {
				return BLZ_ERROR;
			}
			*bs.dst++ = *bs.src++;

			dst_size++;
		}
	}

	/* Return decompressed size */
	return dst_size;
}

/* clang -g -O1 -fsanitize=fuzzer,address -DBLZ_FUZZING depacks.c */
#if defined(BLZ_FUZZING)
#include <limits.h>
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

extern int
LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
{
	if (size > ULONG_MAX / 2) { return 0; }
	void *depacked = malloc(4096);
	if (!depacked) { abort(); }
	blz_depack_safe(data, size, depacked, 4096);
	free(depacked);
	return 0;
}
#endif
